satellite/audit/reservoirservice.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/metainfo"
+)
+
+// ReservoirService is a temp name for the service struct during the audit 2.0 refactor.
+// Once V3-2363 and V3-2364 are implemented, Service2 will replace the existing Service struct.
+type ReservoirService struct {
+	log *zap.Logger
+
+	nodesToSelect  int
+	PathsToAudit   []storj.Path
+	reservoirSlots int
+	Reservoirs     map[storj.NodeID]*Reservoir
+	rand           *rand.Rand
+
+	MetainfoLoop *metainfo.Loop
+	Loop         sync2.Cycle
+}
+
+// NewReservoirService instantiates Service2
+func NewReservoirService(log *zap.Logger, metaLoop *metainfo.Loop, config Config) (*ReservoirService, error) {
+	return &ReservoirService{
+		log: log,
+
+		nodesToSelect:  config.NodesToSelect,
+		reservoirSlots: config.Slots,
+		rand:           rand.New(rand.NewSource(time.Now().Unix())),
+
+		MetainfoLoop: metaLoop,
+		Loop:         *sync2.NewCycle(config.Interval),
+	}, nil
+}
+
+// Run runs auditing service 2.0
+func (service *ReservoirService) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("audit 2.0 is starting up")
+
+	return service.Loop.Run(ctx, func(ctx context.Context) (err error) {
+		defer mon.Task()(&ctx)(&err)
+		pathCollector := NewPathCollector(service.reservoirSlots, service.rand)
+		err = service.MetainfoLoop.Join(ctx, pathCollector)
+		if err != nil {
+			service.log.Error("error joining metainfoloop", zap.Error(err))
+			return nil
+		}
+		service.Reservoirs = pathCollector.Reservoirs
+		return nil
+	})
+}
+
+// Select randomly selects segments to audit
+func (service *ReservoirService) Select(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	return service.Loop.Run(ctx, func(ctx context.Context) (err error) {
+		defer mon.Task()(&ctx)(&err)
+
+		service.PathsToAudit = []storj.Path{}
+		if service.Reservoirs != nil {
+			// todo: is it okay that pathsToAudit could end up being less than nodesToSelect?
+			for i := 0; i < service.nodesToSelect; i++ {
+				randomReservoir, err := service.GetRandomReservoir()
```

Comment:
Why does it need to select randomly if we are going to audit all of the reservoirs anyways?

Reply:
I think according to the design doc we had decided to randomly select reservoirs to choose paths to audit, instead of auditing all reservoirs: https://github.com/storj/storj/blob/master/docs/design/audit-v2.md#design. I thought we were sticking with the idea of auditing as "spot checking" and not thoroughly checking every reservoir. Maybe @jtolds or @wthorp might have more thoughts on this?

Reply:
Yeah, if I recall correctly, we had a meeting where this question came up. The general idea was that random selection is easier to implement and should give more reliable behavior. In the long term, I do not think there is a guarantee that we will try to audit all of the reservoirs, and auditing them in a specific order could result in us being unexpectedly biased. This bias is not present with random selection.